core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -2380,6 +2388,22 @@ private void finish()
             stage.finish();
             taskSource.close();
 
+            sourceStageExecutions.forEach(StageExecution::markFinished);
+
+            recordFinishStats();
+        }
+
+        // parent already finished so no need to continue execution
+        private void markFinished()
```

Comment:
I see some duplication here with the above, but I'm not familiar with the structure of this to suggest a better approach. I'm assuming you considered this.

Reply:
Yeah - I can parametrize. Or maybe name differently. The major difference is that standard `finish` assumes that for each partition processed by the stage there was a task which completed successfully. The `markFinished` is when there is still work to be done by tasks from that stage, but we no longer care.

Reply:
Restructured a bit - hopefully more readable, and less duplication